File: Sources/DebridStreamer/Services/AI/AnthropicProvider.swift
import Foundation

struct AnthropicProvider: AIAssistantProvider {
    let kind: AIProviderKind = .anthropic

    private let apiKey: String
    private let model: String
    private let session: URLSession

    init(apiKey: String, model: String = "claude-3-5-haiku-latest", session: URLSession = .shared) {
        self.apiKey = apiKey
        self.model = model
        self.session = session
    }

    func recommend(prompt: String, candidateTitles: [String], maxResults: Int) async throws -> AIProviderRecommendationResult {
        let trimmedKey = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedKey.isEmpty else {
            throw AIAssistantProviderError.missingAPIKey
        }

        guard let url = URL(string: "https://api.anthropic.com/v1/messages") else {
            throw URLError(.badURL)
        }

        let envelope = AIAssistantJSONParser.promptEnvelope(
            userPrompt: prompt,
            candidateTitles: candidateTitles,
            maxResults: maxResults
        )

        let payload = AnthropicRequest(
            model: model,
            maxTokens: 900,
            temperature: 0.4,
            messages: [.init(role: "user", content: envelope)]
        )

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 45
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(trimmedKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        request.httpBody = try JSONEncoder().encode(payload)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw AIAssistantProviderError.invalidResponse
        }
        guard (200...299).contains(http.statusCode) else {
            let errorText = String(data: data, encoding: .utf8) ?? "Anthropic error"
            throw AIAssistantProviderError.apiError(errorText)
        }

        let decoded = try JSONDecoder().decode(AnthropicResponse.self, from: data)
        let text = decoded.content
            .first(where: { $0.type == "text" })?
            .text
        guard let text else {
            throw AIAssistantProviderError.invalidResponse
        }
        let recommendations = AIAssistantJSONParser.parseRecommendations(from: text, maxResults: maxResults)
        let usage = AIUsageMetrics(
            inputTokens: decoded.usage?.inputTokens,
            outputTokens: decoded.usage?.outputTokens,
            totalTokens: [decoded.usage?.inputTokens, decoded.usage?.outputTokens].compactMap { $0 }.reduce(0, +),
            estimatedCostUSD: AIUsageCostEstimator.estimateUSD(
                model: decoded.model ?? model,
                inputTokens: decoded.usage?.inputTokens,
                outputTokens: decoded.usage?.outputTokens,
                totalTokens: [decoded.usage?.inputTokens, decoded.usage?.outputTokens].compactMap { $0 }.reduce(0, +)
            )
        )
        return AIProviderRecommendationResult(
            model: decoded.model ?? model,
            recommendations: recommendations,
            rawText: text,
            usage: usage
        )
    }
}

private struct AnthropicRequest: Encodable {
    struct Message: Encodable {
        let role: String
        let content: String
    }

    let model: String
    let maxTokens: Int
    let temperature: Double
    let messages: [Message]

    enum CodingKeys: String, CodingKey {
        case model
        case maxTokens = "max_tokens"
        case temperature
        case messages
    }
}

private struct AnthropicResponse: Decodable {
    struct Usage: Decodable {
        let inputTokens: Int?
        let outputTokens: Int?

        enum CodingKeys: String, CodingKey {
            case inputTokens = "input_tokens"
            case outputTokens = "output_tokens"
        }
    }

    struct ContentPart: Decodable {
        let type: String
        let text: String?
    }

    let model: String?
    let content: [ContentPart]
    let usage: Usage?
}

File: web/src/services/ai/AnthropicProvider.ts
// Port of Sources/DebridStreamer/Services/AI/AnthropicProvider.swift.
//
// A fetch-based Anthropic Messages provider. Mirrors the Swift struct: the same
// endpoint/headers/request shape (note `temperature` is intentionally omitted —
// it is rejected by current Claude models), the first text content-part as the
// response text, the usage decoding, and the cost estimate. The `fetch`
// implementation is injectable so tests stub the network (Swift injects a
// URLSession instead).

import type {
  AIProviderKind,
  AIProviderRecommendationResult,
  AIUsageMetrics,
} from "./models";
import {
  AIAssistantJSONParser,
  AIAssistantProviderError,
  AIUsageCostEstimator,
  type AIAssistantProvider,
  type FetchImpl,
  resolveFetch,
  sumTokens,
} from "./types";

const ANTHROPIC_URL = "https://api.anthropic.com/v1/messages";

// MARK: - Request shape (mirrors AnthropicRequest)

interface AnthropicRequestBody {
  model: string;
  max_tokens: number;
  messages: { role: string; content: string }[];
}

// MARK: - Raw response (mirrors AnthropicResponse, snake_case as the API returns)

interface RawAnthropicUsage {
  input_tokens?: number | null;
  output_tokens?: number | null;
}

interface RawAnthropicContentPart {
  type: string;
  text?: string | null;
}

interface RawAnthropicResponse {
  model?: string | null;
  content: RawAnthropicContentPart[];
  usage?: RawAnthropicUsage | null;
}

export class AnthropicProvider implements AIAssistantProvider {
  readonly kind: AIProviderKind = "anthropic";

  private readonly apiKey: string;
  private readonly model: string;
  private readonly fetchImpl: FetchImpl;

  constructor(
    apiKey: string,
    model = "claude-haiku-4-5",
    fetchImpl?: FetchImpl,
  ) {
    this.apiKey = apiKey;
    this.model = model;
    this.fetchImpl = resolveFetch(fetchImpl);
  }

  async recommend(
    prompt: string,
    candidateTitles: string[],
    maxResults: number,
  ): Promise<AIProviderRecommendationResult> {
    const trimmedKey = this.apiKey.trim();
    if (trimmedKey.length === 0) {
      throw AIAssistantProviderError.missingAPIKey();
    }

    const envelope = AIAssistantJSONParser.promptEnvelope(
      prompt,
      candidateTitles,
      maxResults,
    );

    const payload: AnthropicRequestBody = {
      model: this.model,
      max_tokens: 900,
      messages: [{ role: "user", content: envelope }],
    };

    const response = await this.fetchImpl(ANTHROPIC_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-api-key": trimmedKey,
        "anthropic-version": "2023-06-01",
      },
      body: JSON.stringify(payload),
    });

    if (!(response.status >= 200 && response.status <= 299)) {
      const errorText = (await response.text().catch(() => "")) || "Anthropic error";
      throw AIAssistantProviderError.apiError(errorText);
    }

    const decoded = JSON.parse(await response.text()) as RawAnthropicResponse;
    const text = decoded.content.find((part) => part.type === "text")?.text;
    if (text == null) {
      throw AIAssistantProviderError.invalidResponse();
    }

    const recommendations = AIAssistantJSONParser.parseRecommendations(
      text,
      maxResults,
    );

    const resolvedModel = decoded.model ?? this.model;
    const inputTokens = decoded.usage?.input_tokens ?? null;
    const outputTokens = decoded.usage?.output_tokens ?? null;
    const totalTokens = sumTokens(inputTokens, outputTokens);
    const usage: AIUsageMetrics = {
      inputTokens,
      outputTokens,
      totalTokens,
      estimatedCostUSD: AIUsageCostEstimator.estimateUSD(
        resolvedModel,
        inputTokens,
        outputTokens,
        totalTokens,
      ),
    };

    return {
      model: resolvedModel,
      recommendations,
      rawText: text,
      usage,
    };
  }
}
